### DISCUSSION TITLE
How to have an multi-valued option with auto generated help?

### DISCUSSION THREAD

**Discussion by VorpalBlade** [pacak/bpaf#197] [Q&A] [answered]
I want something like this in the help message:

```
my-command [-t] <other args and stuff here>

...
    -t, --template STYLE    Style of template (a (default), b or c)
...
```

Where I have the following code:

```rust
enum Style {
    A,
    B,
    C,
}

impl FromStr for Style {
    // ...
}

#[derive(Debug, Clone, Bpaf)]
#[bpaf(options, version)]
struct Args {
    // ...
    /// Style of template
    #[bpaf(short, long, fallback(Style::A))]
    template: Style,
    // ...
}
```

This clearly doesn't work as is, (the Bpaf macro cannot inspect Style for what styles are supported), I couldn't get any version of this working. I tried adding the Bpaf macro to Style as well and using `bpaf(external)`, but that resulted in `[-a | -b | -c]` which is different from what I want (i.e. I tried https://docs.rs/bpaf/0.7.10/bpaf/params/index.html#required-flag and various variations of that).

I would also like support for completion for this Style enum. Again a custom one could be coded by hand.

Thus my question is: Is there an elegant/automatic way of doing it that I'm missing? Or should I do the complicated hand written one (where I have to keep multiple different lists of things in sync)? And should I file a feature request?

**Comment by pacak** [comment, marked as answer]
It should be possible to implement as combinatoric function similar to [here](https://github.com/pacak/bpaf/blob/07ff03a9850112c6d352f89a8534053d7e9f8691/examples/enum_in_args.rs).

You define a parser for template field something like this:

```rust
fn template() -> impl Parser<Style> {
    let help_msg = format!("....."); // <- you can use `strum` to get exact names and put it in the help message in a way you like

    // again, with help of strum you can get a list of values and decide if it maches
    fn complete_fn(input: &String) -> Vec<...> {}

    // name
    short('t').long("template")
        .help(help_msg) // help mesage generated by strum
        .argument::<String>("ARG")  // consume it as string because completion will need a string to work with
        .complete(complete_fn) // try to complete with help of strum
        .parse(|x| x.parse()) // use `FromStr` to parse it to `Style`
}
```

Then use it in your parser via `external`.

I can try to make a working example if it's still unclear.

**Reply by VorpalBlade** [reply]
I believe I get it, haven't used strum before, but it looks like an interesting crate. (I just hoped for a more streamlined option I guess.)

EDIT: Actually there is one thing I'm not clear on. How do I get the doc strings from my enum for the completion? I use zsh, and it will display such detailed help strings if given the chance.

EDIT 2: Managed to figure that out. However the tab completion still doesn't quite work:

```rust
#[derive(
    Debug, Eq, PartialEq, EnumString, Clone, Copy, EnumIter, EnumMessage, Display, IntoStaticStr,
)]
pub(crate) enum Style {
    /// Normal non-templated file, assumes this program is in PATH
    #[strum(serialize = "normal")]
    Normal,
    /// Templated file, for when this program is in .utils of chezmoi source state
    #[strum(serialize = "template")]
    Template,
}

fn style() -> impl Parser<Style> {
    const DEFAULT: Style = Style::Normal;
    let iter = Style::iter().map(|x| -> String {
        if x == DEFAULT {
            format!("{} (default)", x)
        } else {
            x.to_string()
        }
    });
    let help_msg = format!(
        "How to call the modify manager in the generated file [{}]",
        Itertools::intersperse(iter, ", ".to_string()).collect::<String>()
    );

    fn complete_fn(input: &String) -> Vec<(&'static str, Option<&'static str>)> {
        dbg!(input);
        let res = Style::iter()
            .map(|x| {
                (
                    <Style as Into<&'static str>>::into(x),
                    x.get_documentation(),
                )
            })
            .filter(|(name, _)| name.starts_with(input))
            .collect();
        dbg!(res)
    }

    short('t')
        .long("style")
        .help(help_msg)
        .argument::<String>("STYLE")
        .complete(complete_fn)
        .parse(|x| x.parse())
        .fallback(DEFAULT)
}

#[derive(Debug, Bpaf)]
#[bpaf(options, version)]
pub(super) enum Args {
    /// Process a single file (containing settings).
    Process(#[bpaf(positional("FILE"), complete_shell(ShellComp::File{mask: None}))] PathBuf),
    Add {
        /// Add a file to be tracked by chezmoi_mm
        #[bpaf(short('a'), long("add"))]
        _a: (),
        #[bpaf(external)]
        style: Style,
        #[bpaf(positional("FILE"), complete_shell(ShellComp::File{mask: None}))]
        files: Vec<PathBuf>,
    },
    Smart {
        /// Smartly add a file to be tracked by either chezmoi or chezmoi_mm
        #[bpaf(short('s'), long("smart-add"))]
        _a: (),
        #[bpaf(external)]
        style: Style,
        #[bpaf(positional("FILE"), complete_shell(ShellComp::File{mask: None}))]
        files: Vec<PathBuf>,
    },
    #[cfg(feature = "updater")]
    Update {
        /// Perform self update
        #[bpaf(short('u'), long("update"))]
        _a: (),
    },
}
```

The debug outputs show that my completion code is being executed and returning the results I expect when I e.g. execute: 
```console
$ cargo run -- --bpaf-complete-rev=4  -a -t ''
[chezmoi_mm/src/arguments.rs:28] input = ""
[chezmoi_mm/src/arguments.rs:38] res = [
    (
        "normal",
        Some(
            "Normal non-templated file, assumes this program is in PATH",
        ),
    ),
    (
        "template",
        Some(
            "Templated file, for when this program is in .utils of chezmoi source state",
        ),
    ),
]
```

However, those completions never make it to stdout for some reason, while they do work as expected for completing normal arguments. Also the error messages are cryptic for when the user gets it wrong:

```console
$ cargo run -- -a -t wrongarg file
Expected an argument, got -a
```

Note: This may seem like an overkill solution for something that only take two possible values (and it is). But I have more upcoming options with a similar construction, and those take far more alternatives. Plus I hope to extract the whole thing into a declarative macro to solve it generally for all of my enums.

**Reply by VorpalBlade** [reply]
Hi, sorry to bug you, but do you have any suggestions what could be going on with the weird completion and error reporting issues above?

**Reply by pacak** [reply]
Will take a look and try to reply either later today or tomorrow. Unfortunately github is not sending notifications for edits so I haven't seen your questions.

**Reply by VorpalBlade** [reply]
Oh, that is good to know for the future. Thanks!

**Reply by pacak** [reply]
Do you have above code somewhere in a branch so I can experiment with it?

**Reply by VorpalBlade** [reply]
> Do you have above code somewhere in a branch so I can experiment with it?

Yes, the v2 branch of https://github.com/VorpalBlade/chezmoi_modify_manager/tree/v2 has this. (Not all features do anything yet, as it is a work in progress)

**Reply by pacak** [reply]
Okay, thanks for the example. This was a bug in my code, I'm fixing it in https://github.com/pacak/bpaf/pull/201

If your app is not yet ready to be released to users - you can try running of master branch and somewhere next week I'll try to publish either 0.7.11 with this bug fixed or 0.8 with a bit more changes. If it's urgent - I'll make 0.7.11 with just this bugfix backported.

Also if you are doing something interesting with bpaf like what you are doing in chezmoi and something is not quite working - don't hesitate to make a ticket. While I'm using bpaf completions myself on a daily basis - I'm not covering all the combinations. Here the problem was caused by `fallback` function not propagating collected values.

**Reply by VorpalBlade** [reply]
> If your app is not yet ready to be released to users - you can try running of master branch and somewhere next week I'll try to publish either 0.7.11 with this bug fixed or 0.8 with a bit more changes. If it's urgent - I'll make 0.7.11 with just this bugfix backported.

Not urgent for me, I expect it to be a week or two before the v2 branch is ready anyways. Don't worry.

> Also if you are doing something interesting with bpaf like what you are doing in chezmoi and something is not quite working - don't hesitate to make a ticket.

So people doing boring stuff shouldn't file tickets? ;-)

Seriously though, thanks for taking the time to look into this!

**Reply by pacak** [reply]
> I expect it to be a week or two before the v2 branch is ready anyways.

Okay, I should be done by then.

> So people doing boring stuff shouldn't file tickets? ;-)

Well, I would help either way but I'm assuming boring stuff to be mostly working as expected so just poking for a bit more in boring cases would yield the answer faster.

**Comment by pacak** [comment]
Took a bit longer but changes to generate better help are in master branch, will release soon-ish after going over documentation.
